src/toil_scripts/adam_uberscript/adam_uberscript.py

```diff
@@ -0,0 +1,409 @@
+    """
+    grow the cluster by n nodes
+    """
+    old_size = get_cluster_size(cluster_name)
+    new_size = old_size
+    while diff < n:
+        curr_size = new_size
+        ec2 = connect_to_region(aws_region)
+        zone = { z: ec2.get_spot_price_history(instance_type=instance_type, 
+                                               availability_zone=z,
+                                               product_description='Linux/UNIX',
+                                               start_time=datetime.utcnow().isoformat() )
+                 for z in ( aws_region + c for c in 'ab' ) }
+        log.info('Growing cluster by {} nodes of type: {}'.format(n, instance_type))
+        nodes = subprocess.check_call(['cgcloud',
+                                       'grow-cluster',
+                                       '--list',
+                                       '--instance-type', instance_type,
+                                       '--num-workers', str(n),
+                                       '--cluster-name', cluster_name,
+                                       '--spot-bid', str(spot_bid),
+                                       '--zone', zone,
+                                       'toil'])
+        instance_log.write("Growing cluster. Current instances are: ")
+        for line in nodes:
+            instance_log.write(line)
+        new_size = nodes.count('\n') - 1 # Subtract 1 for header
+        diff = new_size-old_size
+        log.info('Successfully grew cluster by {} nodes of type: {}'.format(new_size-curr_size, instance_type))
+    
+
+def manage_metrics_and_cluster_scaling(params):
+    conn = boto.sdb.connect_to_region(aws_region)
+    dom = conn.get_domain("{0}--files".format(params.jobstore))
+    grow_cluster_thread = threading.Thread(target=monitor_cluster_size, args=(params, conn, dom))
+    metric_collection_thread = threading.Thread(target=collect_realtime_metrics, args=(params, conn, dom))
+    grow_cluster_thread.start()
+    metric_collection_thread.start()
+    grow_cluster_thread.join()
+    metric_collection_thread.join()
+
+def monitor_cluster_size(params, conn, dom):
+    """
+    Monitors cluster size and grows it if the desired size is larger than the current size
+    """
+    log.info('Cluster size monitor has started.')
+    time.sleep(scaling_initial_wait_period_in_seconds)
+    while True:
+        size_check_time = time.time()
+        # If cluster is too small, grow it
+        cluster_size = get_cluster_size()
+        desired_cluster_size = get_desired_cluster_size()
+        if cluster_size < desired_cluster_size:
+            with node_allocation_lock:
+                grow_cluster(desired_cluster_size - cluster_size, params.instance_type, cluster_name, spot_bid)
+        update_cluster_size(desired_cluster_size)
+
+        # Sleep
+        resize_time = time.time() - size_check_time
+        log.info('Cluster is {} nodes as of {}'.format(get_cluster_size(), resize_time))
```

Comment:
Small logging irregularity here. I've opened #121 to track this; it is not important to fix now, but we should fix it later.